File: blanc/src/algorithm/leetcode/p201to250/LeetCode201BitwiseAndOfNumbersRange.java
package algorithm.leetcode.p201to250;

import org.junit.Assert;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;

public class LeetCode201BitwiseAndOfNumbersRange {

    private interface Method {
        int rangeBitwiseAnd(int m, int n);
    }

    private static class Comparing implements Method {

        @Override
        public int rangeBitwiseAnd(int m, int n) {
            if (m == 0) {
                return 0;
            }
            int res = 0;
            int cursor = 1 << 30;
            while (cursor > n) {
                cursor >>= 1;
            }
            while ((cursor & m) == (cursor & n) && cursor > 0) {
                res += (cursor & m);
                cursor >>= 1;
            }
            return res;
        }
    }

    private static class Move implements Method {

        @Override
        public int rangeBitwiseAnd(int m, int n) {
            if (m == 0) {
                return 0;
            }
            int v1 = m;
            int v2 = n;
            int shift = 0;
            while (v1 != v2 && v1 != 0) {
                v1 >>= 1;
                v2 >>= 1;
                shift++;
            }
            return v1 << shift;
        }
    }

    private static Method getMethod() {
        return new Comparing();
    }

    @Test
    public void testcase1() {
        test(5, 7, 4);
    }

    private void test(int m, int n, int expect) {
        Assert.assertThat(getMethod().rangeBitwiseAnd(m, n), is(expect));
    }
}

File: src/leetcode/p201to250/LeetCode201BitwiseAndOfNumbersRange.kt
package leetcode.p201to250

class LeetCode201BitwiseAndOfNumbersRange {
    interface Method {
        fun rangeBitwiseAnd(m: Int, n: Int): Int
    }

    class MyMethod : Method {

        override fun rangeBitwiseAnd(m: Int, n: Int): Int {
            if (m == 0) {
                return 0
            }
            var cursor = 1 shl 30
            var res = 0
            while (cursor > n) {
                cursor = cursor.shr(1)
            }
            while (cursor.and(m) == cursor.and(n) && cursor > 0) {
                res += cursor.and(m)
                cursor = cursor.shr(1)
            }
            return res
        }
    }

    class RefMethod : Method {

        override fun rangeBitwiseAnd(m: Int, n: Int): Int {
            if (m == 0) {
                return 0
            }
            var cursor = 1
            var v1 = m
            var v2 = n
            while (v1 != v2) {
                v1 = v1.shr(1)
                v2 = v2.shr(1)
                cursor = cursor.shl(1)
            }
            return v1 * cursor
        }
    }
}